src/main/java/org/truffleruby/core/module/ModuleOperations.java

```diff
@@ -393,36 +392,41 @@ public static ConstantLookupResult lookupConstantWithInherit(RubyContext context
         return definedMethods;
     }
 
-    @TruffleBoundary
     public static MethodLookupResult lookupMethodCached(DynamicObject module, String name,
             DeclarationContext declarationContext) {
+        return lookupMethodCached(module, null, name, declarationContext);
+    }
+
+    @TruffleBoundary
+    private static MethodLookupResult lookupMethodCached(DynamicObject module, DynamicObject lookupTo, String name,
+            DeclarationContext declarationContext) {
         final ArrayList<Assumption> assumptions = new ArrayList<>();
 
-        // Look in ancestors
         for (DynamicObject ancestor : Layouts.MODULE.getFields(module).ancestors()) {
-            final ModuleFields fields = Layouts.MODULE.getFields(ancestor);
-            assumptions.add(fields.getMethodsUnmodifiedAssumption());
-            final InternalMethod method = fields.getMethod(name);
-            if (method != null) {
-                if (method.isRefined()) {
-                    if (declarationContext != null) {
-                        final DynamicObject[] refinements = declarationContext.getRefinementsFor(ancestor);
-                        if (refinements != null) {
-                            for (DynamicObject refinement : refinements) {
-                                final MethodLookupResult refinedMethod = lookupMethodCached(refinement, name, null);
-                                if (refinedMethod.isDefined()) {
-                                    for (Assumption assumption : refinedMethod.getAssumptions()) {
-                                        assumptions.add(assumption);
-                                    }
-                                    return new MethodLookupResult(refinedMethod.getMethod(), toArray(assumptions));
-                                }
-                            }
+            if (ancestor == lookupTo) {
+                return new MethodLookupResult(null, toArray(assumptions));
+            }
+            final DynamicObject[] refinements = getRefinementsFor(declarationContext, ancestor);
+            if (refinements != null) {
+                for (int i = 0; i < refinements.length; i++) {
+                    final MethodLookupResult refinedMethod = lookupMethodCached(
+                            refinements[i],
+                            getLookupTo(refinements, ancestor, i),
+                            name,
+                            null);
+                    if (refinedMethod.isDefined()) {
+                        for (Assumption assumption : refinedMethod.getAssumptions()) {
+                            assumptions.add(assumption);
                         }
```

Comment:
These 3 lines should go above the `if`.
So we're notified if any of the modules looked up has new/changed/removed methods.

Reply:
Is this valid for `lookupSuperMethod` too?

Reply:
I would think so. Whether we found a result or not, we want to know if there is a method added earlier in the ancestors later on.